---
title: "Psi1 snRNAseq"
subtitle: "Pass 2: DE and QC"
author: "Kennedi Todd"
date: "05/27/2025"
output:
  html_document:
  code_folding: hide
theme: cerulean
toc: true
toc_float: true
editor_options: 
  chunk_output_type: inline
---

# Setup
## Working directory
```{r working_directory}
knitr::opts_knit$set(root.dir = ".")
```

## Libraries
```{r libraries, message=FALSE, warning=FALSE}
# load packages
#library(ComplexUpset) # instersection_size()
library(DESeq2)       # DESeqDataSetFromMatrix
library(dotenv)       # load_dot_env()
library(dplyr)        # distinct()
library(ggplot2)      # scale_colour_gradientn()
library(ggrepel)      # geom_text_repel()
#library(gridExtra)   # grid.arrange()
library(gtools)       # smartbind()
#library(parallel)    # detectCores()
#library(plotly)      # plot_ly()
library(Seurat)       # DimPlot()
library(stringr)      # str_match
library(tidyr)        # %>%
#library(UpSetR)      # fromList()
```

## Variables and functions
```{r set_variables_and_functions}
# source thresholds and output paths
source("../../refs/thresholds_and_outs.R")
out <- paste0(out, "pass2/")

# single cell functions
files <- list.files("../../functions", full.names = TRUE)
invisible(lapply(files, source))
```

## Load data
```{r read_annotated_obj}
mouse.annotated <- readRDS(paste0("../../rObjects/", filtering_method, 
                                  "_pass2_annotated_seurat_obj.rds"))
```

## UMAP
```{r annotated_umap}
cluster_colors <- get_cluster_colors(
  num_clusters = length(unique(mouse.annotated$annotated_clusters))
)

DimPlot(
  object = mouse.annotated,
  cols = cluster_colors
)
```

# MAST
```{r}
# done in another script and submitted to HPC
```

# QC
## Markers by sample
```{r}
canonical_markers <- c(
  "Aldoc", "Aqp4", "Gja1", "Clu", "Apoe",             # Astrocytes
  "Cfap44", "Spag16", "Ttr", "Folr1",                 # Choroid
  "Ptprb", "Cldn5", "Flt1",                           # Endothelial
  "Pdgfrb", "Rgs5", "Acta2", "Vtn",                   # Mural
  "Col1a1", "Col1a2", "Dcn",                          # Fibroblast
  "Aif1", "C1qa", "P2ry12", "Tmem119","Trem2","Mrc1", # Microglia
  "Syt1", "Snap25", "Gad1", "Meg3", "Igfbpl1","Sox4", # Neurons
  "Pdgfra", "Cspg4", "Olig1", "Olig2", "Tnr",         # OPCs
  "Mog", "Opalin", "Mal", "Mag", "Mbp",               # Oligodendrocytes
  "Cd3e","Itgax","S100a9"                             # Immune (T, DC, neutrophil)
)

d <- DotPlot(mouse.annotated,
             features = canonical_markers,
             group.by = "annotated_clusters",
             split.by = "sample_id",
             cols = "RdYlBu",
             assay = "RNA") +
  coord_flip() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) 
d
```

```{r}
pdf(paste0(out, "markers/annotated_canonical_markers_split_sample.pdf"), 
    width = 25, height = 12)
d
dev.off()
remove(d)
```

## Heatmap UMAP
```{r umap_heatmap, warning=FALSE, message=FALSE}
# UMAP percent_mt
f1 <- FeaturePlot(mouse.annotated, 
            reduction = "umap", 
            features = "percent_mt")  + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f1

# UMAP nCount
f2 <- FeaturePlot(mouse.annotated, 
            reduction = "umap",
            features = "nCount_RNA") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f2

# UMAP nFeature
f3 <- FeaturePlot(mouse.annotated, 
            reduction = "umap", 
            features = "nFeature_RNA") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f3

# UMAP percent_ribo
f4 <- FeaturePlot(mouse.annotated, 
            reduction = "umap", 
            features = "percent_ribo") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f4

# UMAP cell_complexity
f5 <- FeaturePlot(mouse.annotated, 
            reduction = "umap", 
            features = "cell_complexity") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f5

# UMAP percent_hb
f6 <- FeaturePlot(mouse.annotated, 
            reduction = "umap", 
            features = "percent_hb") + 
  scale_colour_gradientn(colours = c("blue","lightblue","yellow","orange","red"))
f6
```

```{r save_umap_heatmap, echo=FALSE, eval=FALSE}
path <- paste0(out, "clustering_QC/umap_heatmap_")
pdf(paste0(path, "percent_mito.pdf"), width = 8, height = 6)
f1
dev.off()

pdf(paste0(path, "nCount.pdf"), width = 8, height = 6)
f2
dev.off()

pdf(paste0(path, "nFeature.pdf"), width = 8, height = 6)
f3
dev.off()

pdf(paste0(path, "percent_ribo.pdf"), width = 8, height = 6)
f4
dev.off()

pdf(paste0(path, "cell_complexity.pdf"), width = 8, height = 6)
f5
dev.off()

pdf(paste0(path, "percent_hb.pdf"), width = 8, height = 6)
f6
dev.off()

remove(f1,f2,f3,f4,f5,f6)
```

## Percent cells per cluster
```{r percent_cells}
# timepoint
b1 <- mouse.annotated@meta.data %>%
  group_by(annotated_clusters, timepoint) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=timepoint)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of timepoint per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b1

# sex
b2 <- mouse.annotated@meta.data %>%
  group_by(annotated_clusters, sex) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=sex)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of sex per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b2

# sample
b3 <- mouse.annotated@meta.data %>%
  group_by(annotated_clusters, sample_id) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=sample_id)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of sample per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b3

# phase
b4 <- mouse.annotated@meta.data %>%
  group_by(annotated_clusters, phase) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=phase)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of phase per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b4

# mito.factor
b5 <- mouse.annotated@meta.data %>%
  group_by(annotated_clusters, mito_factor) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=mito_factor)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of mito factor per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b5

# group
b6 <- mouse.annotated@meta.data %>%
  group_by(annotated_clusters, group) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=group)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of group per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b6

# group2
b7 <- mouse.annotated@meta.data %>%
  group_by(annotated_clusters, group2) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=group2)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of group2 per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b7

# dose
b8 <- mouse.annotated@meta.data %>%
  group_by(annotated_clusters, dose) %>%
  dplyr::count() %>%
  group_by(annotated_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=annotated_clusters,y=percent, fill=dose)) +
  theme_classic() +
  geom_col() +
  ggtitle("Percentage of dose per cluster") +
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust=1))
b8
```

```{r save_percent_cells,echo=FALSE,eval=FALSE}
# save
path <- paste0(out, "clustering_QC/percent_")
pdf(paste0(path, "timepoint_per_cluster.pdf"), width = 8, height = 6)
b1
dev.off()

# save
pdf(paste0(path, "sex_per_cluster.pdf"), width = 8, height = 6)
b2
dev.off()

# save
pdf(paste0(path, "sample_per_cluster.pdf"), width = 8, height = 8)
b3
dev.off()

# save
pdf(paste0(path, "phase_per_cluster.pdf"), width = 8, height = 6)
b4
dev.off()

# save
pdf(paste0(path, "mito_factor_per_cluster.pdf"), width = 8, height = 6)
b5
dev.off()

# save
pdf(paste0(path, "group_per_cluster.pdf"), width = 8, height = 6)
b6
dev.off()

# save
pdf(paste0(path, "group2_per_cluster.pdf"), width = 8, height = 6)
b7
dev.off()

# save
pdf(paste0(path, "age_dose_cluster.pdf"), width = 8, height = 6)
b8
dev.off()

# cleanup
remove(b1,b2,b3,b4,b5,b6,b7,b8)
```

## Cells per sample
```{r cells_per_sample}
data <- as.data.frame(table(mouse.annotated$sample_id))
colnames(data) <- c("sample_id","frequency")

ncells <- ggplot(data, aes(x = sample_id, y = frequency, fill = sample_id)) + 
  geom_col() +
  theme_classic() +
  geom_text(aes(label = frequency), 
            position=position_dodge(width=0.9), 
            vjust=-0.25) +
  scale_y_continuous(breaks = seq(0,25000, by = 5000), limits = c(0,25000)) +
  ggtitle("Filtered: cells per sample") +
  theme(legend.position =  "none") + 
  theme(axis.text.x = element_text(angle = 45, hjust=1))
ncells
```

```{r save_cells_per_sample, echo=FALSE, eval=FALSE}
pdf(paste0(out, "clustering_QC/cells_per_sample.pdf"), width = 10, height = 4)
ncells
dev.off()
remove(ncells)
```

## Cells per cluster
```{r cells_per_cluster}
data <- as.data.frame(table(mouse.annotated$annotated_clusters))
colnames(data) <- c("cluster","frequency")

ncells <- ggplot(data, aes(x = cluster, y = frequency, fill = cluster)) + 
  geom_col() +
  theme_classic() +
  geom_text(aes(label = frequency), 
            position=position_dodge(width=0.9), 
            vjust=-0.25) +
  scale_y_continuous(breaks = seq(0,150000, by = 25000), limits = c(0,150000)) +
  ggtitle("Cells per Cluster") +
  scale_fill_manual(values = cluster_colors) +
  theme(legend.position =  "none") + 
  theme(axis.text.x = element_text(angle = 45, hjust=1))
ncells
```

```{r save_cells_per_cluster, echo=FALSE, eval=FALSE}
pdf(paste0(out, "clustering_QC/cells_per_cluster.pdf"), width = 8, height = 6)
ncells
dev.off()
remove(data,ncells)
```

## Cells per cluster per sample
```{r cells_per_cluster_per_sample}
# extract data
data <- data.frame(cluster = mouse.annotated$annotated_clusters,
                   sample = mouse.annotated$sample_id)
data_summary <- data %>%
  group_by(cluster, sample) %>%
  summarise(count = n(), .groups = 'drop') %>%
  ungroup()

# Create the stacked bar plot
ncells <- ggplot(data_summary, aes(x = cluster, y = count, fill = sample)) +
  geom_bar(stat = "identity") +
  labs(y = "Frequency", x = "Cluster", fill = "Sample") +
  theme_minimal() +
  ggtitle("Frequency of Cells in Each Cluster with Sample Distribution") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust=1)) +
  ylim(c(0,150000))
ncells

# reformat for table output
data_summary <- data_summary %>% pivot_wider(names_from = cluster, values_from = count)
```

```{r save_cells_per_cluster_per_sample, echo=FALSE, eval=FALSE}
pdf(paste0(out, "clustering_QC/cells_per_cluster_with_sample_distribution.pdf"),
    width = 8, height = 6)
ncells
dev.off()
remove(ncells,data,data_summary)
```

## Cells per cluster per group
```{r cells_per_cluster_per_group}
# extract data
data <- data.frame(cluster = mouse.annotated$annotated_clusters,
                   group = mouse.annotated$group)
data_summary <- data %>%
  group_by(cluster, group) %>%
  summarise(count = n(), .groups = 'drop') %>%
  ungroup()

# Create the stacked bar plot
ncells <- ggplot(data_summary, aes(x = cluster, y = count, fill = group)) +
  geom_bar(stat = "identity") +
  labs(y = "Frequency", x = "Cluster", fill = "Group") +
  theme_minimal() +
  ggtitle("Frequency of Cells in Each Cluster with Group Distribution") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust=1)) +
  ylim(c(0,150000))
ncells

# reformat for table output
data_summary <- data_summary %>% pivot_wider(names_from = cluster, values_from = count)
```

```{r save_cells_per_cluster_per_group, echo=FALSE, eval=FALSE}
pdf(paste0(out, "clustering_QC/cells_per_cluster_with_group_distribution.pdf"),
    width = 8, height = 6)
ncells
dev.off()

write.table(data_summary, 
            file = paste0(out, "clustering_QC/cells_per_cluster_with_group_distribution.tsv"), 
            sep = "\t",quote = FALSE, row.names = FALSE)
```

## Sex check
```{r sex_check}
dot <- DotPlot(mouse.annotated,
               features = c("Uty","Xist"),
               group.by = "sample_id")
dot
```

```{r save_sex_check, echo=FALSE, eval=FALSE}
pdf(paste0(out, "clustering_QC/sex_check_dot_plot.pdf"), width = 6, height = 8)
dot
dev.off()
remove(dot)
```

# DEG Plots
## Compare DEGs
### All 
```{r compare_DEGs}
# set thresh
adjpval_thresh <- 0.05
lfc_thresh <- 0
pct_thresh <- 0.6

# get file paths
files <- list.files(paste0(out, "DEGs_MAST_pct_", pct_thresh, "/DEG_tables"))
files <- files[grepl(".+_DEGs.tsv", files)]

# init data.frame
deg.df <- data.frame()

# loop through files
for (i in files) {
  
  # read table
  df <- read.table(paste0(out, "DEGs_MAST_pct_", pct_thresh, "/DEG_tables/", i),
                   sep = "\t", 
                   header = TRUE)
  
  # filter
  df <- df[!is.na(df$p_val_adj),]
  df <- df[df$p_val_adj < adjpval_thresh,]
  df <- df[df$avg_log2FC < -lfc_thresh | df$avg_log2FC > lfc_thresh,]
  
  # add direction column
  comparisonName <- gsub("_DEGs.tsv", "", i)
  direction <- df$avg_log2FC > 0 # already filtered on lfc so only matters it's positive/negative
  direction <- gsub(TRUE, paste0(comparisonName, "_up"), direction)
  direction <- gsub(FALSE, paste0(comparisonName, "_down"), direction)
  df$direction <- direction
  
  # reformat table
  df <- df %>% 
    dplyr::count(cluster, direction) %>%
    tidyr::spread(cluster, n)
  
  # add to MASTer table
  deg.df <- smartbind(deg.df, df)
}

# reorder rows
rownames(deg.df) <- 1:nrow(deg.df)
deg.df <- deg.df[c(7:12,1:6),]
rownames(deg.df) <- deg.df$direction
deg.df$direction <- NULL

# reformat
deg.df[is.na(deg.df)] <- 0

# save
prefix <- paste0(out, "DEGs_MAST_pct_", pct_thresh, "/DEG_comparison/",
                 "all_comparisons_FDRq_", format(adjpval_thresh, nsmall = 2),
                 "_LFC_", format(lfc_thresh, nsmall = 2))
write.table(deg.df, 
            paste0(prefix, ".tsv"),
            sep = "\t", quote = FALSE, row.names = TRUE)

# set heatmap colors and names
paletteLength <- 100
myColor <- colorRampPalette(c("white","#f0eb9e","darkgreen"))(paletteLength)

meta <- data.frame(treatment = c(rep("low dose psilocybin", 6),
                                 rep("high dose psilocybin", 6)),
                   sex = c("both sexes", "both sexes",
                           "female", "female",
                           "male", "male",
                           "both sexes", "both sexes",
                           "female", "female",
                           "male", "male"),
                   direction = c(rep(c("down-regulated","up-regulated"), 6))
                   )
rownames(meta) <- rownames(deg.df)

ann_colors <- list(treatment = c(`low dose psilocybin` = "green", 
                                 `high dose psilocybin` = "purple"),
                   sex = c(`female` = "gold", 
                           `male` = "gray",
                           `both sexes` = "cyan"),
                   direction = c(`down-regulated` = "#0271f0", 
                                 `up-regulated` = "tomato"))

# save
pdf(paste0(prefix, ".pdf"), width = 10, height = 6)

# plot
pheatmap::pheatmap(deg.df,
                   main = paste0("FDRq < ", format(adjpval_thresh, nsmall = 2), 
                                 ", |LFC| > ", format(lfc_thresh, nsmall = 2)),
                   treeheight_row = 0,
                   treeheight_col = 0,
                   color = myColor,
                   cluster_rows = FALSE,
                   annotation_row = meta,
                   annotation_colors = ann_colors,
                   display_numbers = round(deg.df, digits = 0),
                   fontsize_number = 12,
                   number_color = "black")

# cleanup
remove(ann_colors, deg.df, df, meta)
```

### Sex specific
```{r compare_sex_specific_DEGs, eval=FALSE}
# set thresh
adjpval_thresh <- 0.05
lfc_thresh <- 0
pct_thresh <- 0.6

# get file paths
files <- list.files(paste0(out, "DEGs_MAST_pct_", pct_thresh, "/DEG_tables"))
files <- files[grepl(".+[MF]_vs_.+[MF]_DEGs.tsv", files)]

# init data.frame
deg.df <- data.frame()

# loop through files
for (i in files) {
  
  # read table
  df <- read.table(paste0(out, "DEGs_MAST_pct_", pct_thresh, "/DEG_tables/", i),
                   sep = "\t", 
                   header = TRUE)
  
  # filter
  df <- df[!is.na(df$p_val_adj),]
  df <- df[df$p_val_adj < adjpval_thresh,]
  df <- df[df$avg_log2FC < -lfc_thresh | df$avg_log2FC > lfc_thresh,]
  
  # add direction column
  comparisonName <- gsub("_DEGs.tsv", "", i)
  direction <- df$avg_log2FC > 0 # already filtered on lfc so only matters it's positive/negative
  direction <- gsub(TRUE, paste0(comparisonName, "_up"), direction)
  direction <- gsub(FALSE, paste0(comparisonName, "_down"), direction)
  df$direction <- direction
  
  # reformat table
  df <- df %>% 
    dplyr::count(cluster, direction) %>%
    tidyr::spread(cluster, n)
  
  # add to MASTer table
  deg.df <- smartbind(deg.df, df)
}

# reformat
rownames(deg.df) <- 1:nrow(deg.df)
deg.df[is.na(deg.df)] <- 0
deg.df <- deg.df[c(5:8,1:4),]
rownames(deg.df) <- deg.df$direction
deg.df$direction <- NULL

# save
prefix <- paste0(out, "DEGs_MAST_pct_", pct_thresh, "/DEG_comparison/",
                 "sex_specific_comparison_FDRq_", format(adjpval_thresh, nsmall = 2),
                 "_LFC_", format(lfc_thresh, nsmall = 2))
write.table(deg.df, 
            paste0(prefix, ".tsv"),
            sep = "\t", quote = FALSE, row.names = TRUE)

# set heatmap colors and names
paletteLength <- 100
myColor <- colorRampPalette(c("white","#f0eb9e","darkgreen"))(paletteLength)

meta <- data.frame(treatment = c(rep("Low psilocybin", 4),
                                 rep("High psilocybin", 4)),
                   sex = c(rep("Female", 2), rep("Male", 2),
                           rep("Female", 2), rep("Male", 2)),
                   direction = rep(c("down-regulated","up-regulated"), 4))
rownames(meta) <- rownames(deg.df)

ann_colors <- list(treatment = c(`Low psilocybin` = "green", `High psilocybin` = "purple"),
                   sex = c(`Female` = "gold", `Male` = "gray"),
                   direction = c(`down-regulated` = "#0271f0", `up-regulated` = "tomato"))

# save
pdf(paste0(prefix, ".pdf"), width = 10, height = 6)

# plot
pheatmap::pheatmap(deg.df,
                   main = paste0("FDRq < ", format(adjpval_thresh, nsmall = 2), 
                                 ", |LFC| > ", format(lfc_thresh, nsmall = 2)),
                   treeheight_row = 0,
                   treeheight_col = 0,
                   color = myColor,
                   cluster_rows = FALSE,
                   annotation_row = meta,
                   annotation_colors = ann_colors,
                   display_numbers = round(deg.df, digits = 0),
                   fontsize_number = 12,
                   number_color = "black")

remove(ann_colors, deg.df, df, meta)
```

### Both sexes
```{r compare_both_sexes_DEGs}
# set thresh
adjpval_thresh <- 0.05
lfc_thresh <- 0
pct_thresh <- 0.2

# get file paths
files <- list.files(paste0(out, "DEGs_MAST_pct_", pct_thresh, "/DEG_tables"))
files <- files[grepl("[LH].24h_vs_S.24h_DEGs.tsv", files)]

# init data.frame
deg.df <- data.frame()

# loop through files
for (i in files) {
  
  # read table
  df <- read.table(paste0(out, "DEGs_MAST_pct_", pct_thresh, "/DEG_tables/", i),
                   sep = "\t", 
                   header = TRUE)
  
  # filter
  df <- df[!is.na(df$p_val_adj),]
  df <- df[df$p_val_adj < adjpval_thresh,]
  df <- df[df$avg_log2FC < -lfc_thresh | df$avg_log2FC > lfc_thresh,]
  
  # add direction column
  comparisonName <- gsub("_DEGs.tsv", "", i)
  direction <- df$avg_log2FC > 0 # already filtered on lfc so only matters it's positive/negative
  direction <- gsub(TRUE, paste0(comparisonName, "_up"), direction)
  direction <- gsub(FALSE, paste0(comparisonName, "_down"), direction)
  df$direction <- direction
  
  # reformat table
  df <- df %>% 
    dplyr::count(cluster, direction) %>%
    tidyr::spread(cluster, n)
  
  # add to master table
  deg.df <- smartbind(deg.df, df)
}

# reformat
deg.df[is.na(deg.df)] <- 0
rownames(deg.df) <- deg.df$direction
deg.df$direction <- NULL

# reorder rows
deg.df <- deg.df[c(3:4,1:2),]

# save
prefix <- paste0(out, "DEGs_MAST_pct_", pct_thresh, "/DEG_comparison/",
                 "both_sexes_DEG_comparison_FDRq_", format(adjpval_thresh, nsmall = 2),
                 "_LFC_", format(lfc_thresh, nsmall = 2))
write.table(deg.df, 
            paste0(prefix, ".tsv"),
            sep = "\t", quote = FALSE, row.names = TRUE)

# set heatmap colors and names
paletteLength <- 100
myColor <- colorRampPalette(c("white","#f0eb9e","darkgreen"))(paletteLength)

meta <- data.frame(treatment = c(rep("Low psilocybin", 2),
                                 rep("High psilocybin", 2)),
                   direction = rep(c("down-regulated","up-regulated"), 2))
rownames(meta) <- rownames(deg.df)

ann_colors <- list(treatment = c(`Low psilocybin` = "green", `High psilocybin` = "purple"),
                   direction = c(`down-regulated` = "#0271f0", `up-regulated` = "tomato"))

# save
pdf(paste0(prefix, ".pdf"), width = 10, height = 6)

# plot
pheatmap::pheatmap(deg.df,
                   main = paste0("FDRq < ", format(adjpval_thresh, nsmall = 2), 
                                 ", |LFC| > ", format(lfc_thresh, nsmall = 2)),
                   treeheight_row = 0,
                   treeheight_col = 0,
                   color = myColor,
                   cluster_rows = FALSE,
                   annotation_row = meta,
                   annotation_colors = ann_colors,
                   display_numbers = round(deg.df, digits = 0),
                   fontsize_number = 12,
                   number_color = "black")

remove(ann_colors, deg.df, df, meta)
```

## Volcano 
```{r volcano, message=FALSE, eval=FALSE}
# set thresh
adjpval_thresh <- 0.05
lfc_thresh <- 0
pct_thresh <- 0.6

# get file paths
files <- list.files(paste0(out, "DEGs_MAST_pct_", pct_thresh, "/DEG_tables"), full.names = TRUE)
files <- files[grepl(".+_DEGs.tsv", files)]

for (file in files) {
  
  # read DEG file
  treatment_vs_control <- read.delim(file)
  
  # filter NA values
  treatment_vs_control <- treatment_vs_control[!is.na(treatment_vs_control$p_val_adj),]
  
  # extract title for plot later
  p.comparison <- str_match(file, "DEG_tables/(.+)_DEGs.tsv")[,2]
  
  # assign colors
  color_values <- vector()
  max <- nrow(treatment_vs_control)
  for(row in 1:max){
    if (treatment_vs_control$p_val_adj[row] < adjpval_thresh){
      if (treatment_vs_control$avg_log2FC [row] > lfc_thresh){
        color_values <- c(color_values, 1) # 1 when logFC > thresh and FDRq < thresh
      } else if (treatment_vs_control$avg_log2FC[row] < -lfc_thresh){
        color_values <- c(color_values, 2) # 2 when logFC < thresh and FDRq < thresh
      } else {
        color_values <- c(color_values, 3) # 2 when FDRq < thresh and LFC not met
      }
    }
    else{
      color_values <- c(color_values, 3) # 3 when FDRq >= thresh
    }
  }
  treatment_vs_control$color_adjpval_thresh <- factor(color_values)
  
  # loop through clusters
  for (j in unique(treatment_vs_control$cluster)) {
    
    # subset cluster
    data <- subset(treatment_vs_control, cluster == j)
    
    # plot only if there are DEGs with p_val_adj < thresh
    num <- subset(data, p_val_adj < adjpval_thresh)
    num <- nrow(num)
    if(num != 0) {
        
      # subset genes to label
      up <- data[data$color_adjpval_thresh == 1,]
      up.sig <- up[order(up$p_val_adj),][1:15,]
      up.lfc <- up[order(up$avg_log2FC, decreasing = TRUE),][1:15,]
      up30 <- rbind(up.sig,up.lfc)
      up30 <- unique(up30)
      up30 <- up30[rowSums(is.na(up30)) != ncol(up30), ]
      down <- data[data$color_adjpval_thresh == 2,]
      down.sig <- down[order(down$p_val_adj),][1:15,]
      down.lfc <- down[order(down$avg_log2FC, decreasing = FALSE),][1:15,]
      down30 <- rbind(down.sig,down.lfc)
      down30 <- unique(down30)
      down30 <- down30[rowSums(is.na(down30)) != ncol(down30), ]
      
      # set manual colors
      if (!1 %in% unique(data$color_adjpval_thresh)) {
        my_colors <- c("blue","gray")
      } else if (!2 %in% unique(data$color_adjpval_thresh)) {
        my_colors <- c("red","gray")
      } else if (!1 %in% unique(data$color_adjpval_thresh) && !2 %in% unique(data$color_adjpval_thresh)) {
        my_colors <- c("gray")
      } else {
        my_colors <- c("red","blue","gray")
      }
      
      # set significance threshold
      hadjpval <- (-log10(max(
        data$p_val[data$p_val_adj < adjpval_thresh], 
        na.rm=TRUE)))

      # plot
      p <-
        ggplot(data = data, 
               aes(x = avg_log2FC,     # x-axis is logFC
                   y = -log10(p_val),  # y-axis will be -log10 of P.Value
                   color = color_adjpval_thresh)) +  # color is based on factored color column
        geom_point(alpha = 0.8, size = 2) +  # create scatterplot, alpha makes points transparent
        theme_bw() +  # set color theme
        theme(legend.position = "none") +  # no legend
        scale_color_manual(values = my_colors) +
        labs(
          title = "",
          x = expression(log[2](FC)),
          y = expression(-log[10] ~ "(" ~ italic("p") ~ "-value)")
        ) +
        theme(axis.title.x = element_text(size = 10),
              axis.text.x = element_text(size = 10)) +
        theme(axis.title.y = element_text(size = 10),
              axis.text.y = element_text(size = 10)) +
        geom_hline(yintercept = hadjpval, colour = "#000000", linetype = "dashed") +
        geom_vline(xintercept = lfc_thresh, colour = "#000000", linetype = "dashed") +
        geom_vline(xintercept = -lfc_thresh, colour = "#000000", linetype = "dashed") +
        ggtitle(paste0(j, "\n", p.comparison, ", q < ", format(adjpval_thresh, nsmall = 2), 
                       ", |LFC| > ", format(lfc_thresh, nsmall = 2))) +
        geom_text_repel(data = up30,
                        aes(x = avg_log2FC, y= -log10(p_val), label = gene), 
                        color = "maroon", 
                        fontface="italic",
                        max.overlaps = getOption("ggrepel.max.overlaps", default = 30)
                        ) +
        geom_text_repel(data = down30,
                        aes(x = avg_log2FC, y= -log10(p_val), label = gene), 
                        color = "navyblue", 
                        fontface="italic",
                        max.overlaps = getOption("ggrepel.max.overlaps", default = 30)
                        )
      p
      
      # create dir if doesn't exist
      prefix <- paste0(out, "DEGs_MAST_pct_", pct_thresh, "/volcano/", p.comparison)
      if (!dir.exists(prefix)) {
        dir.create(prefix, recursive = TRUE)
      }
      
      # save
      j_replace <- gsub(" ", "_", tolower(j))
      j_replace <- gsub("/", "_", j_replace)
      plot_out <- paste0(prefix, "/", p.comparison, "_", j_replace,
                    "_FDRq_", format(adjpval_thresh, nsmall = 2), "_LFC_", 
                    format(lfc_thresh, nsmall = 2), "_volcano.pdf")
      pdf(plot_out, height = 8, width = 8)
      print(p)
      dev.off()
      
      print(paste("comparison =",p.comparison,", cell type =", j))
      
    }
  } # end loop through clusters
} # end loop through variables
```

## Metascape input
```{r metascape_sex_specific}
# set thresh
adjpval_thresh <- 0.05
lfc_thresh <- 0
pct_thresh <- 0.2

# get file paths
files <- list.files(paste0(out, "DEGs_MAST_pct_", pct_thresh, "/DEG_tables"), full.names = TRUE)
files <- files[grepl(".+_DEGs.tsv", files)]

# loop through files
for (i in 1:length(files)) {
  
  # read table
  data <- read.table(files[i], header = TRUE, sep = "\t")
  
  # filter
  data <- data[!is.na(data$p_val_adj),]
  data <- data[data$p_val_adj < adjpval_thresh,]
  data <- data[data$avg_log2FC < -lfc_thresh | data$avg_log2FC > lfc_thresh,]
  
  # separate into up and down-regulated
  up.df <- data[data$avg_log2FC > 0,]
  down.df <- data[data$avg_log2FC < 0,]
  
  cell_types <- unique(data$cluster)
  
  # create dir if doesn't exist
  p.comparison <- str_match(files[i], "DEG_tables/(.+)_DEGs.tsv")[,2]
  prefix <- paste0(out, "DEGs_MAST_pct_", pct_thresh, "/metascape_input/", p.comparison)
  if (!dir.exists(prefix)) {
    dir.create(prefix, recursive = TRUE)
  }
  
  # loop through cell_type
  for (j in cell_types) {
    
    # subset based on cell_type
    print(paste0(p.comparison, ": ", j))
    up <- up.df[up.df$cluster == j,]
    up <- up$gene
    down <- down.df[down.df$cluster == j,]
    down <- down$gene
    
    # file name
    prefix <- paste0(out, "DEGs_MAST_pct_", pct_thresh, "/metascape_input/", p.comparison,
                     "/", p.comparison, "_", gsub(" ", "_", tolower(j)), 
                     "_FDRq_", format(adjpval_thresh, nsmall = 2), 
                     "_LFC_", format(lfc_thresh, nsmall = 2))
     
    # save
    if(length(up) > 10) {
      write.table(x = up,
                  file = paste0(prefix, "_up.tsv"),
                  quote = FALSE,
                  row.names = FALSE,
                  col.names = FALSE)      
    }
    if (length(down) > 10) {
       write.table(x = down,
                file = paste0(prefix, "_down.tsv"),
                quote = FALSE,
                row.names = FALSE,
                col.names = FALSE)
    }
  } # end j for loop
} # end i for loop
```

# Shiny App
- scDRcell is the function used to plot the first UMAP
- inpConf is an argument of scDRcell, inpConf = sc1conf
- infConf gets subset by UI column based on inpdrX which is the user's input for x-axis
- inpdrX is used to set the x-axis label
```{r shiny_app, eval=FALSE}
# create new object
library(ShinyCell)
shiny.obj <- mouse.annotated
VariableFeatures(shiny.obj) <- shiny.obj@assays$SCT@var.features

# set default params
DefaultAssay(shiny.obj) <- "RNA"

# create config
names <- colnames(shiny.obj@meta.data)
names <- names[c(31,32,2:27)]
sc.config <- createConfig(obj = shiny.obj,
                          meta.to.include = names)

# change wd
setwd(out)

# output shiny app folder
makeShinyApp(obj = shiny.obj, 
             scConf = sc.config, 
             gene.mapping = TRUE,
             shiny.title = "E. coli Mice")

# manual config edits
setwd(out)
sc1conf <- readRDS("shinyApp/sc1conf.rds")
cluster_hex <- "#fa4e3e|#fabf3e|#55f531|#08780c|#78f3fa|#2966f2|#b186f7|#fa78e9"
sc1conf[2,4] <- cluster_hex
saveRDS(sc1conf, "shinyApp/sc1conf.rds")

# cleanup
remove(sc.config,sc1conf,shiny.obj)
```

# doubletFinder preparation
```{r doubletFiner_split_save}
# Split by sample
split <- SplitObject(mouse.annotated, split.by = "sample")

# Save each sample as its own .rds file
for (i in seq_along(split)) {
  sample_name <- names(split)[i]
  save_path <- paste0("../../rObjects/split_seurat_objects/", sample_name, ".rds")
  saveRDS(split[[i]], file = save_path, compress = FALSE)
  cat("Saved:", save_path, "\n")
}
```
