---
title: "Psi1 snRNAseq"
subtitle: "Postfiltering QC"
author: "Kennedi Todd"
date: "05/21/2025"
output:
  html_document:
    code_folding: hide
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---

# Setup
## Set working directory
```{r working_directory}
knitr::opts_knit$set(root.dir = ".")
```

## Load libraries
```{r libraries, message=FALSE, warnings=FALSE}
# load libraries
library(cowplot)     # plot_grid()
library(dplyr)       # left_join()
library(ggplot2)     # ggplot()
library(gridExtra)   # grid.arrange()
library(rtracklayer) # import()
library(scCustomize) # Merge_Seurat_List()
library(Seurat)      # Read10X_h5()
library(stringr)     # str_match()
```

## Variables and functions
```{r set_variables_and_thresholds}
# source thresholds and output paths
source("../../refs/thresholds_and_outs.R")
out <- paste0(out, "pass1/")
```

## Load data
```{r load_data}
mouse.filtered <- readRDS(paste0( "../../rObjects/", filtering_method, 
                                  "_pass1_quality_filtered_seurat_obj.rds"))
```

# Post-filtering QC
## Number of cells
```{r postfiltering_ncells}
data <- as.data.frame(table(mouse.filtered$sample_id))
colnames(data) <- c("sample_id","frequency")

ncells2 <- ggplot(data, aes(x = sample_id, y = frequency, fill = sample_id)) + 
  geom_col() +
  theme_bw() +
  geom_text(aes(label = frequency), 
            position=position_dodge(width=0.9), 
            vjust=-0.25) +
  #scale_fill_manual(values = sample_colors) + 
  scale_y_continuous(breaks = seq(0,40000, by = 10000), limits = c(0,40000)) +
  ggtitle("Filtered: cells per sample") +
  theme(legend.position =  "none") + 
  theme(axis.text.x = element_text(angle = 45, hjust=1))
ncells2
```

```{r save_cells_per_sample}
# save
pdf(paste0(out, "filtering_QC/ncells_filtered.pdf"),
    width = 18, height = 4)
ncells2

# cleanup
remove(ncells2)
```

## Density plots
```{r postfiltering_density, warning=FALSE}
# Visualize nCount_RNA
den1 <- ggplot(mouse.filtered@meta.data,
       aes(color = sample_id,
           x = nCount_RNA,
           fill = sample_id)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  xlab("nCount_RNA") +
  ylab("Density") +
  theme(legend.position =  "none") +
  geom_vline(xintercept = nCount.min, lty = "dashed") +
  geom_vline(xintercept = nCount.max, lty = "dashed") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9)) +
  scale_x_log10() +
  annotate("text", x = nCount.min, y = 0.1, label = paste("Min:", nCount.min), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3) +
  annotate("text", x = nCount.max, y = 0.1, label = paste("Max:", nCount.max), 
           angle = 0, vjust = -0.5, hjust = 1, size = 3)

# Visualize nFeature_RNA
den2 <- ggplot(mouse.filtered@meta.data,
       aes(color = sample_id,
           x = nFeature_RNA,
           fill = sample_id)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  theme(legend.position =  "none") +
  xlab("nFeature_RNA") +
  ylab("Density") +
  geom_vline(xintercept = nFeature.min, lty = "dashed") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9)) +
  annotate("text", x = nFeature.min, y = 0.1, label = paste("Min:", nFeature.min), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)

# Visualize cell complexity
den3 <- ggplot(mouse.filtered@meta.data,
       aes(color = sample_id,
           x = cell_complexity,
           fill = sample_id)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  #scale_color_manual(values = sample_colors) +
  theme(legend.position =  "none") +
  #scale_fill_manual(values = sample_colors) +
  xlab("Cell Complexity (log10(nFeature/nCount))") +
  ylab("Density") +
  geom_vline(xintercept = complexity.cutoff, lty = "dashed") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9)) +
  annotate("text", x = complexity.cutoff, y = 0.1, label = paste("Min:", complexity.cutoff), 
           angle = 0, vjust = -0.5, hjust = -0.1, size = 3)

# Visualize percent_mt
den4 <- ggplot(mouse.filtered@meta.data,
       aes(color = sample_id,
           x = percent_mt,
           fill = sample_id)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_continuous(n.breaks = 4) +
  geom_vline(xintercept = mt.cutoff, lty = "dashed") +
  theme(legend.position =  "none") +
  xlab("% Mitochondrial Genes") +
  ylab("Density") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9)) +
  annotate("text", x = mt.cutoff, y = 0.1, label = paste("Max:", mt.cutoff), 
           angle = 0, vjust = -0.5, hjust = 1, size = 3)

# Visualize percent_ribo
den5 <- ggplot(mouse.filtered@meta.data,
       aes(color = sample_id,
           x = percent_ribo,
           fill = sample_id)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  theme(legend.position =  "none") +
  xlab("% Ribosomal Protein Genes") +
  ylab("Density") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9))

# Visualize percent_hb
den6 <- ggplot(mouse.filtered@meta.data,
       aes(color = sample_id,
           x = percent_hb,
           fill = sample_id)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  geom_vline(xintercept = hb.cutoff, lty = "dashed") +
  theme(legend.position =  "none") +
  xlab("% Hemoglobin Genes") +
  ylab("Density") +
  geom_vline(xintercept = hb.cutoff, lty = "dashed") +
  theme(legend.key.size = unit(0.25, 'cm'), legend.title = element_text(size=9)) +
  annotate("text", x = hb.cutoff, y = 0.1, label = paste("Max:", hb.cutoff), 
           angle = 0, vjust = -0.5, hjust = 1, size = 3)

# Arrange graphs in grid
plots <- list(den1,den2,den3,den4,den5,den6)
layout <- rbind(c(1,4),c(2,5),c(3,6))
grid <- grid.arrange(grobs = plots, layout_matrix = layout)
```

```{r save_postfiltering_density, warning=FALSE}
# save
pdf(paste0(out, "filtering_QC/density_curves_filtered.pdf"), width = 8, height = 8)
grid <- grid.arrange(grobs = plots, layout_matrix = layout)

# cleanup
remove(den1,den2,den3,den4,den5,den6,plots,layout,grid)
```

## Violin plots
```{r postfiltering_violins}
# nFeature, nCount, and cell.complexity violins
v3 <- VlnPlot(mouse.filtered,
              features = c("nFeature_RNA", "nCount_RNA","cell_complexity"),
              ncol = 1,
              group.by = 'sample_id',
              raster = FALSE,
              pt.size = 0)
v3

#  percent violins
v4 <- VlnPlot(mouse.filtered,
              features = c("percent_mt","percent_ribo","percent_hb"),
              ncol = 1,
              raster = FALSE,
              group.by = 'sample_id',
              pt.size = 0)
v4
```

```{r save_postfiltering_violins}
# save v3
pdf(paste0(out, "filtering_QC/violins_nFeature_nCount_complexity_filtered.pdf"), 
    width = 12, height = 12)
v3
dev.off()

# save v4
pdf(paste0(out, "filtering_QC/violins_percent_filtered.pdf"),
    width = 12, height = 12)
v4
dev.off()

# cleanup
remove(v3,v4)
```

## Scatter plots
```{r postfiltering_scatter, warning=FALSE}
s4 <- ggplot(
  mouse.filtered@meta.data,
  aes(x = nCount_RNA, y = nFeature_RNA, color = percent_mt)) + 
  geom_point() + 
  geom_smooth(method = "lm") +
	scale_x_log10() +   	
  scale_y_log10() + 
  theme_classic() +
  geom_vline(xintercept = nCount.min, lty = "dashed", col = "red") + 
  geom_hline(yintercept = nFeature.min, lty = "dashed", col = "red") + 
  geom_vline(xintercept = nCount.max, lty = "dashed", col = "red") +
  facet_wrap(~sample_id) +
  scale_colour_gradient(low = "gray90", high = "black", limits =c(0,100))
s4

s5 <- FeatureScatter(mouse.filtered,
                     feature1 = "nCount_RNA",
                     feature2 = "percent_mt",
                     group.by = 'sample_id',
                     shuffle = TRUE)
s5

s6 <- ggplot(
  mouse.filtered@meta.data,
  aes(x = nCount_RNA, y = nFeature_RNA, color = percent_hb)) + 
  geom_point() + 
  geom_smooth(method = "lm") +
	scale_x_log10() +   	
  scale_y_log10() + 
  theme_classic() +
  geom_vline(xintercept = nCount.min, lty = "dashed", col = "red") + 
  geom_hline(yintercept = nFeature.min, lty = "dashed", col = "red") + 
  facet_wrap(~sample_id) +
  scale_colour_gradient(low = "gray90", high = "black", limits =c(0,100))
s6
```

```{r save_postfiltering_scatter}
# save
path <- paste0(out, "filtering_QC/scatter_nFeature_vs_nCount_filtered.pdf")
pdf(path, width = 18, height = 18)
s4
dev.off()

path <- paste0(out, "filtering_QC/scatter_nCount_vs_percentMT_filtered.pdf")
pdf(path, width = 10, height = 6)
s5
dev.off()

path <- paste0(out, "filtering_QC/scatter_nCount_vs_percentMT_filtered2.pdf")
pdf(path, width = 18, height = 18)
s6
dev.off()

# cleanup
remove(s4,s5,s6)
```

## Box plot
```{r boxplot}
# Visualize the distribution of genes detected per cell via boxplot
b1 <- ggplot(mouse.filtered@meta.data,
       aes(x = sample_id, 
           y = log10(nFeature_RNA), 
           fill=sample_id)) + 
  geom_boxplot() + 
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust = 0.5, face="bold")) +
  ggtitle("Unique Genes / Cell / Sample") +
  xlab("Sample")
b1
```

```{r save_boxplot}
# save
path <- paste0(out, "filtering_QC/boxplot_nFeature_per_sample.pdf")
pdf(path, width = 12, height = 6)
b1
dev.off()

# cleanup
remove(b1)
```

## Top transcripts
```{r top_transcripts}
df <- data.frame(gene_name = rownames(mouse.filtered))
df$rsum <- rowSums(x = mouse.filtered, slot = "counts")
df <- df[order(df$rsum, decreasing = TRUE),]
rownames(df) <- 1:nrow(df)
head(df, 10)
```

```{r save_top_transcripts}
write.table(df, 
            paste0(out, "filtering_QC/top_transcripts.tsv"),
            quote = FALSE,
            row.names = TRUE)
```

