---
title: "Psilocybin Project 1"
subtitle: "Sleuth QC"
author: "Kennedi Todd"
date: "11/19/2025"
output:
  html_document:
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---

# Background
- bulk RNA sequencing of prefrontal cortex of 90 mice \
- 5 samples per group (treatment, dose, time point, sex) \
- high dose = 1 mg/kg, low dose = 0.25 mg/kg

# Setup
```{r setup, include=FALSE}
# Set working directory globally
knitr::opts_knit$set(root.dir = ".")
```

## Libraries
```{r libraries, message=FALSE, warning=FALSE}
library(dotenv)  # load_dot_env
library(dplyr)   # %>%
library(ggplot2) # ggplot
library(ggrepel) # geom_text_repel
library(gtools)  # smartbind
library(sleuth)  # sleuth_prep
library(tidyr)   # pivot_wider
```

## User defined variables
```{r set_variables}
# Get output directories and file paths
out <- "../results/sex_specific/"
degs_dir <- paste0(out, "DEGs/")
model <- "group2_hbb"

# Load environment variables
load_dot_env(file = "../refs/.env")  # Adjust the path to your .env file

# Colors and levels
treatment_order <- c("saline", "psilocybin")
treatment_colors <- c("gray60", "springgreen2")

sex_order <- c("M", "F")
sex_colors <- c("gold2", "purple")

group_order <- c("S.8h", "S.24h", "S.7d", 
                 "L.8h", "L.24h", "L.7d", 
                 "H.8h", "H.24h", "H.7d")
group_colors <- c("gray90", "gray60", "gray30", 
                  "lightblue", "cornflowerblue", "blue", 
                  "sienna1", "red2", "red4")
```

## Read sleuth object
```{r}
so <- readRDS(paste0("../rObjects/so_", model, ".rds"))
```

# QC
## PCA
```{r plot_pca}
p1 <- plot_pca(so, color_by = "group2")
p1

p2 <- plot_pca(so, color_by = "group")
p2

p3 <- plot_pca(so, color_by = "sex")
p3

p4 <- plot_pca(so, color_by = "treatment")
p4

p5 <- plot_pca(so, color_by = "dose")
p5

p6 <- plot_pca(so, color_by = "timepoint")
p6
```

```{r save_pca}
pdf(paste0(out, "PCA/PCA_colored_by_group2.pdf"), width = 8, height = 6)
p1
dev.off()

pdf(paste0(out, "PCA/PCA_colored_by_group.pdf"), width = 8, height = 6)
p2
dev.off()

pdf(paste0(out, "PCA/PCA_colored_by_sex.pdf"), width = 8, height = 6)
p3
dev.off()

pdf(paste0(out, "PCA/PCA_colored_by_treatment.pdf"), width = 8, height = 6)
p4
dev.off()

pdf(paste0(out, "PCA/PCA_colored_by_dose.pdf"), width = 8, height = 6)
p5
dev.off()

pdf(paste0(out, "PCA/PCA_colored_by_time_point.pdf"), width = 8, height = 6)
p6
dev.off()

remove(p1,p2,p3,p4,p5,p6)
```

## Sample density
```{r}
plot_sample_density(so)
```

## Sample heatmap
```{r}
plot_sample_heatmap(so)
```

## Scatter
```{r}
plot_scatter(so)
```

```{r}
plot_fld(so, sample = "Psi1_A1_High_Male_L3_kallisto")
```

# Get Hbb-bs TPM
```{r}
hbb_transcripts <- subset(so$target_mapping, gene_name == "Hbb-bs")
hbb_transcripts <- hbb_transcripts[,"target_id"]

tpm <- so$obs_norm_filt

hbb_tpm <- subset(tpm, target_id %in% hbb_transcripts) %>% 
  select(sample, tpm)
hbb_tpm$hbb_tpm <- hbb_tpm$tpm
hbb_tpm$tpm <- NULL

meta <- so$sample_to_covariates
meta <- left_join(x = meta, y = hbb_tpm, by = "sample")
so$sample_to_covariates <- meta

remove(hbb_tpm, tpm, meta)
```

# DE plots
## Comparison table
```{r}
qval_thresh <- 0.1
lfc_thresh <- 0
deg_files <- list.files(paste0(degs_dir, "DEG_tables/"))
master_deg <- data.frame()

# loop through files
for (i in deg_files) {
  
  # read table
  df <- read.table(paste0(degs_dir, "DEG_tables/", i), 
                   sep = "\t", header = TRUE)
  
  # filter
  df <- df[!is.na(df$qval),]
  df <- df[df$qval < qval_thresh,]
  df <- df[df$b > lfc_thresh | df$b < -lfc_thresh,]
  
  # add direction column
  comparisonName <- gsub("_DEGs.tsv", "", i)
  
  # get up/down direction
  direction <- df$b > 0
  direction <- gsub(TRUE, "up", direction)
  direction <- gsub(FALSE, "down", direction)
  df$direction <- direction
  
  # reformat table
  df <- df %>% dplyr::count(direction) 
  rownames(df) <- df$direction
  df$direction <- NULL
  df <- as.data.frame(t(df))
  df$comparison <- comparisonName
  
  # add to master table
  master_deg <- smartbind(master_deg, df)
}

# replace NAs
master_deg[is.na(master_deg)] <- 0

# reorder rows
rownames(master_deg) <- 1:nrow(master_deg)
master_deg <- master_deg[c(11,12,7,8,9,10,5,6,1,2,3,4,15,13,14),]

# reformat
rownames(master_deg) <- master_deg$comparison
master_deg$comparison <- NULL

# save
write.table(x = master_deg,
            file = paste0(degs_dir, "DEG_comparison/DEG_comparison_qval_", 
                          format(qval_thresh, nsmall = 2), 
                          "_lfc_", format(lfc_thresh, nsmall = 2),
                          ".tsv"),
            sep = "\t",
            row.names = TRUE,
            col.names = TRUE,
            quote = FALSE)

# cleanup
remove(df)
```

## Comparison heatmap
```{r}
# set heatmap colors and names
meta <- data.frame(treatment = c(rep("low dose", 6),
                                rep("high dose", 6),
                                rep("saline", 3)),
                   sex = c(rep(c("female","male"),6), rep("both",3)),
                   timepoint = c(rep(c("8 hours","8 hours",
                                       "24 hours","24 hours",
                                       "7 days", "7 days"),
                                     2),
                                 "8 hours", 
                                 "24 hours", 
                                 "7 days"))
rownames(meta) <- rownames(master_deg)
paletteLength <- 100
myColor <- colorRampPalette(c("white","#f0eb9e","darkgreen"))(paletteLength)
ann_colors = list(treatment = c(`low dose` = "cornflowerblue", 
                               `high dose` = "firebrick2",
                               `saline` = "gray"),
                  sex = c(female = 'chartreuse2',
                          male = 'chocolate4',
                          both = "pink"),
                  timepoint = c(`8 hours` = "gold",
                                `24 hours` = "purple",
                                `7 days` = "black"))

# save
path <- paste0(degs_dir, "DEG_comparison/DEG_comparison_qval_",
               format(qval_thresh, nsmall = 2), "_lfc_", 
               format(lfc_thresh, nsmall = 2), ".pdf")
pdf(path, width = 8, height = 9)

# plot
pheatmap::pheatmap(master_deg,
                   main = paste0("qval < ", format(qval_thresh, nsmall = 2), 
                                 ", |lfc| > ", format(lfc_thresh, nsmall = 2)),
                   treeheight_row = 0,
                   treeheight_col = 0,
                   color = myColor,
                   cluster_rows = FALSE,
                   annotation_row = meta,
                   annotation_colors = ann_colors,
                   display_numbers = round(master_deg, digits = 0),
                   fontsize_number = 12,
                   number_color = "black")

# cleanup
remove(ann_colors,meta)
```


## Volcano
```{r volcano_plot, message=FALSE, eval=FALSE}
# get file list
qval_thresh <- 0.1
b_thresh <- 0.2
deg_files <- list.files(paste0(degs_dir, "DEG_tables/"))

for (i in deg_files) {
  
  # read DEG file
  data <- read.delim(paste0(degs_dir, "DEG_tables/", i), sep = "\t")
  data <- data[!is.na(data$qval),]
  
  # assign colors
  color_values <- vector()
  max <- nrow(data)
  for(row in 1:max){
    if (data$qval[row] < qval_thresh){
      if (data$b [row] > b_thresh){
        color_values <- c(color_values, 1) # 1 when b > b_thresh and qval < qval_thresh
      } else if (data$b[row] < -b_thresh){
        color_values <- c(color_values, 2) # 2 when b < -b_thresh and qval < qval_thresh
      } else {
        color_values <- c(color_values, 3) # 3 when b between -b_thresh and +b_thresh and qval < qval_thresh
      }
    } else {
      color_values <- c(color_values, 3) # 3 when qval >= qval_thresh
    }
  }
  data$color_qval <- factor(color_values)
  
 # comparison name
 comparison <- gsub(paste0(degs_dir, "DEG_tables/"), "", i)
 comparison <- gsub("_DEGs.tsv","",comparison)
    
 # plot only if there are DEGs with qval < qval_thresh
 num <- data[data$qval < qval_thresh,]
 num <- nrow(num)
 if(num != 0) {
   
    # subset genes to label
    up <- data[data$color_qval == 1,]
    up.sig <- up[order(up$qval),][1:15,]
    up.b <- up[order(up$b, decreasing = TRUE),][1:15,]
    up30 <- rbind(up.sig,up.b)
    up30 <- up30[!duplicated(up30$transcript_name),]
    down <- data[data$color_qval == 2,]
    down.sig <- down[order(down$qval),][1:15,]
    down.b <- down[order(down$b, decreasing = FALSE),][1:15,]
    down30 <- rbind(down.sig,down.b)
    down30 <- down30[!duplicated(down30$transcript_name),]
      
    # set manual colors
    if (!1 %in% unique(data$color_qval)) {
      my_colors <- c("blue","gray")
    } else if (!2 %in% unique(data$color_qval)) {
      my_colors <- c("red","gray")
    } else if (!1 %in% unique(data$color_qval) && !2 %in% unique(data$color_qval)) {
      my_colors <- c("gray")
    } else {
      my_colors <- c("red","blue","gray")
    }
      
    # set significance threshold
    hqval <- (-log10(max(data$pval[data$qval < qval_thresh], na.rm=TRUE)))

    # plot
    p <-
      ggplot(data = data, 
             aes(x = b,  # x-axis is b
                 y = -log10(pval),  # y-axis will be -log10 of pval
                 color = color_qval)) +  # color is based on factored color column
      geom_point(alpha = 0.8, size = 2) +  # create scatterplot, alpha makes points transparent
      theme_bw() +  # set color theme
      theme(legend.position = "none") +  # no legend
      scale_color_manual(values = my_colors) +  # set factor colors
      labs(
        title = "", # no main title
        x = expression(log[2](FC)), # x-axis title
         y = expression(-log[10] ~ "(" ~ italic("p") ~ "-value)") # y-axis title
      ) +
      theme(axis.title.x = element_text(size = 15),
            axis.text.x = element_text(size = 15)) +
      theme(axis.title.y = element_text(size = 15),
            axis.text.y = element_text(size = 15)) +
      theme(plot.title = element_text(size = 15)) +
      geom_hline(yintercept = hqval,  #  horizontal line
                         colour = "#000000",
                         linetype = "dashed") +
      geom_vline(xintercept = -b_thresh,  #  vertical line
                         colour = "#000000",
                         linetype = "dashed") +
      geom_vline(xintercept = b_thresh,  #  vertical line
                         colour = "#000000",
                         linetype = "dashed") +
      ggtitle(paste0(comparison, ", qval < ", qval_thresh, ", b = ", b_thresh)) +
      geom_text_repel(data = up30,
                      aes(x = b, y= -log10(pval), label = transcript_name), 
                      size = 5,
                      color = "maroon", 
                      fontface="italic",
                      max.overlaps = getOption("ggrepel.max.overlaps", default = 30)
                      ) +
      geom_text_repel(data = down30,
                      aes(x = b, y= -log10(pval), label = transcript_name), 
                      color = "navyblue", 
                      size = 5,
                      fontface="italic",
                      max.overlaps = getOption("ggrepel.max.overlaps", default = 30)
                      )
     p
      
    # save
    path <- paste0(degs_dir, "volcano/", comparison, "_qval_", 
                   format(qval_thresh, nsmall = 2), "_b_",
                   format(b_thresh, nsmall = 2), "_volcano.pdf")
    pdf(path, height = 8, width = 8)
    print(p)
    dev.off()
  }
} # end loop through variables
```

## Metascape input
```{r}
# get file list
qval_thresh <- 0.1
b_thresh <- 0.2
deg_files <- list.files(paste0(degs_dir, "DEG_tables/"))

# intitalize table for command line version of metascape
msbio.df <- data.frame()

# loop through DEG files
for (i in 1:length(deg_files)) {
  # read table
  data <- read.table(paste0(degs_dir, "DEG_tables/", deg_files[i]), header = TRUE, sep = "\t")
  data <- data[!is.na(data$qval),]
  
  # filter based on adjusted p-value
  data <- data[data$qval < qval_thresh,]
  
  # create up-regulated gene list
  up <- data[data$b > b_thresh,]
  up <- unique(up$gene_name)
  
  # create down-regulated gene list
  down <- data[data$b < -b_thresh,]
  down <- unique(down$gene_name)
  
 # get the comparison name
 comparison <- deg_files[i]
 comparison <- gsub("_DEGs.tsv", "", comparison)
 
 # make two filenames for up and downregulated genes
 # include the comparison, adjusted p-value threshold, and logFC threshold
 up.filename <- paste0(degs_dir, "metascape_input/", comparison,
                       "_upregulated_qval_", format(qval_thresh, nsmall = 2), 
                       "_b_", format(b_thresh, nsmall = 2), ".txt")
 down.filename <- paste0(degs_dir, "metascape_input/", comparison,
                         "_downregulated_qval_", format(qval_thresh, nsmall = 2), 
                         "_b_", format(b_thresh, nsmall = 2), ".txt")
 
 
  # save the up and down-reulated gene lists
  write.table(x = up,
              file = up.filename,
              quote = FALSE,
              row.names = FALSE,
              col.names = FALSE)
  write.table(x = down,
              file = down.filename,
              quote = FALSE,
              row.names = FALSE,
              col.names = FALSE)
    
  # job for MSBio
  # if there are more than 10 up-regulated genes, add paths in json format for msbio job
  if (length(up) > 10) {
      msbio <- jsonlite::toJSON(list(
        input = gsub(paste0(degs_dir, "metascape_input/"),
                     "/data/metascape_input/",
                     up.filename),
        output = paste0("/data/metascape_output/",
                        comparison,
                        "_upregulated_qval_", 
                        format(qval_thresh, nsmall = 2), 
                        "_b_", format(b_thresh, nsmall = 2)),
        single = TRUE))
      msbio <- gsub("\\[","",msbio)
      msbio <- gsub("\\]","",msbio)
      msbio.df <- rbind(msbio.df, msbio)
  }
  # if there are more than 10 down-regulated genes, add paths in json format for msbio job
  if (length(down) > 10) {
      msbio <- jsonlite::toJSON(list(
        input = gsub(paste0(degs_dir, "metascape_input/"),
                     "/data/metascape_input/",
                     down.filename),
        output = paste0("/data/metascape_output/",
                        comparison,
                        "_downregulated_qval_", 
                        format(qval_thresh, nsmall = 2),
                        "_b_", 
                        format(b_thresh, nsmall = 2)),
        single = TRUE))
      msbio <- gsub("\\[","",msbio)
      msbio <- gsub("\\]","",msbio)
      msbio.df <- rbind(msbio.df, msbio)
  }
}

# save msbio job
write.table(x = msbio.df,
            file = paste0(degs_dir, "metascape_input/msbio.job"),
            quote = FALSE,
            row.names = FALSE,
            col.names = FALSE)
```

# Isoform switching
## TPM
```{r}
tpm <- kallisto_table(so)
tpm <- tpm %>% select(target_id, sample, tpm, group2)
```

## PIU
```{r}
# add gene info
tpm <- left_join(x = tpm, 
                 y = so$target_mapping[,c("target_id", 
                                          "transcript_name",
                                          "gene_id", 
                                          "gene_name")],
                  by = "target_id")

# get percentage isoform usage
piu <- tpm %>%
  group_by(gene_id, sample) %>%
  filter(sum(tpm) > 0) %>% 
  mutate(piu = tpm / sum(tpm)) %>%
  ungroup()

# remove genes that only have one isoform in all samples
genes_multi_all <- piu %>%
  group_by(gene_id, sample) %>%
  summarise(n_iso = n_distinct(target_id[tpm > 0])) %>%
  ungroup() %>%
  group_by(gene_id) %>%
  summarise(is_multi_all = all(n_iso > 1)) %>%
  filter(is_multi_all) %>%
  pull(gene_id)

piu_filtered <- piu %>% filter(gene_id %in% genes_multi_all)

# compare
dim(piu)
dim(piu_filtered)
```


## Mean PIU per group
```{r}
piu_means <- piu_filtered %>%
  group_by(gene_id, gene_name, target_id, transcript_name, group2) %>%
  summarise(mean_piu = mean(piu), .groups = "drop")
```

## Change in PIU
```{r}
# contrasts table
contrast_df <- tibble::tibble(
  group1 = c("L.8h.F","L.8h.M","L.24h.F","L.24h.M","L.7d.F","L.7d.M",
             "H.8h.F","H.8h.M","H.24h.F","H.24h.M","H.7d.F","H.7d.M",
             "S.8h.F","S.24h.F","S.7d.F"),
  group2 = c("S.8h.F","S.8h.M","S.24h.F","S.24h.M","S.7d.F","S.7d.M",
             "S.8h.F","S.8h.M","S.24h.F","S.24h.M","S.7d.F","S.7d.M",
             "S.8h.M","S.24h.M","S.7d.M")
)

# function
compute_delta_piu <- function(piu_means, g1, g2) {

  piu_means %>%
    filter(group2 %in% c(g1, g2)) %>%
    select(gene_id, gene_name, target_id, transcript_name, group2, mean_piu) %>%
    tidyr::pivot_wider(
      names_from = group2,
      values_from = mean_piu,
      values_fill = 0
    ) %>%
    mutate(
      delta_piu = !!sym(g1) - !!sym(g2),
      comparison = paste0(g1, "_vs_", g2)
    )
}

delta_piu <- purrr::map_dfr(
  1:nrow(contrast_df),
  ~ compute_delta_piu(
      piu_means,
      contrast_df$group1[.x],
      contrast_df$group2[.x]
    )
)
```


## Global switching
- closer to 1 = this isoform fully dominates the gene in at least one condition \
- closer to 0 = this isoform is never dominant anywhere
- intermediate = isoform is neither always off nor always dominant
```{r}
piu_means <- piu %>%
  group_by(gene_id, target_id, group2) %>%
  summarize(mean_piu = mean(piu), .groups = "drop")

piu_global <- piu_means %>%
  group_by(gene_id, target_id) %>%
  summarize(
    max_piu = max(mean_piu, na.rm = TRUE),
    min_piu = min(mean_piu, na.rm = TRUE),
    range_piu = max(mean_piu, na.rm = TRUE) - min(mean_piu, na.rm = TRUE),
    .groups = "drop"
  )
```

```{r}
# Which condition each isoform 'prefers'
piu_pref <- piu_means %>%
  group_by(gene_id, target_id) %>%
  slice_max(order_by = mean_piu, n = 1, with_ties = FALSE) %>%
  ungroup() %>%
  rename(pref_group = group2)

# Genes where different isoforms prefer different conditions
switch_genes <- piu_pref %>%
  group_by(gene_id) %>%
  filter(n_distinct(pref_group) > 1) %>%
  distinct(gene_id)

```

## DE isoform switches
```{r}
deg_files <- list.files(paste0(out, "DEGs/DEG_tables/"))
deg_files <- deg_files[grep("[LH].+_vs_S.+", deg_files)]
iso_gene_info <- data.frame()

# loop through files
for (i in deg_files) {
  
  # read table
  data <- read.table(paste0(out, "DEGs/DEG_tables/", i), sep = "\t", header = TRUE)
  
  # filter
  data <- data[!is.na(data$b),]
  
  # get comparison name
  comparisonName <- gsub("_DEGs.tsv", "", i)
  
  # get fold change
  direction <- data[,c("target_id","transcript_name","gene_id","gene_name","b")]
  direction[[comparisonName]] <- data$b
  direction$b <- NULL
  
  # add to master table
  if (nrow(iso_gene_info) == 0) {
    iso_gene_info <- direction
  } else (
    iso_gene_info <- left_join(x = iso_gene_info, y = direction)
  )
}

# remove genes that only have one isoform
iso_gene_info <- iso_gene_info %>%
  group_by(gene_id) %>%
  filter(n() > 1) %>%
  ungroup()


# count positives/negatives
iso_gene_info$num_positive <- rowSums(iso_gene_info[,5:ncol(iso_gene_info)] > 0)
iso_gene_info$num_negative <- rowSums(iso_gene_info[,5:ncol(iso_gene_info)] < 0)

# remove transcripts that all trend the same way
iso_gene_info <- iso_gene_info[!iso_gene_info$num_positive == 12,]
iso_gene_info <- iso_gene_info[!iso_gene_info$num_negative == 12,]

# save
write.table(x = master_deg,
            file = paste0(out, "DEGs/DEG_comparison/DEG_comparison_qval_", 
                          format(qval_thresh, nsmall = 2),".tsv"),
            sep = "\t",
            row.names = TRUE,
            col.names = TRUE,
            quote = FALSE)

# cleanup
remove(df)
```

# isoformSwitchAnalyzeR
## Import
```{r}
importIsoformExpression
```

